[Brendon Lau] iP

BrendonLau opened the pull request:
# Duke
> “Your mind is for having ideas, not holding them.” – David Allen([source](https://dansilvestre.com/productivity-quotes/))

Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/BrendonLau/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

BrendonLau committed:
Level-1

BrendonLau committed:
Level-2

Added notebook class for adding of tasks

BrendonLau committed:
Level-3

Added Task class and Command class

BrendonLau committed:
Levell-4

Event, deadline and todo inherits from Task.

BrendonLau committed:
A-TextUiTesting

- Modified runtest.bat
- test inputs all passed

BrendonLau committed:
Error Handling

- Handle commands that are not valid
- Handle tasks that are not valid
- Handle marking/unmarking of invalid task id
- Handle empty task

BrendonLau committed:
added delete feature to TaskManager

- Support the delete of task by using remove() from ArrayList
- Update test cases

BrendonLau committed:
A-Enums

- Enum is used to support querying command for switch.

BrendonLau committed:
Update code format

BrendonLau committed:
Save data from input to text file

Save tasklist into a file.
Load tasklist into the program.

BrendonLau committed:
Update Task to accept datetime format

BrendonLau committed:
Merge branch 'branch-level-8'

BrendonLau committed:
A-MoreOOP

- Added UI
- Added Parser

- Constant for easy reference

BrendonLau committed:
A-Packages

- Added duke package
-Added duke.common and duke.task package

BrendonLau committed:
Merge branch 'add-gradle-support' of https://github.com/BrendonLau/ip""

BrendonLau committed:
A-JUnit

JUnit testing for 3 classes

BrendonLau committed:
A-Jar

BrendonLau committed:
A-JavaDoc

Added javadocs to most of the non-private classes and methods except the trivial ones.

BrendonLau committed:
A-CodingStandard

BrendonLau committed:
Update feature to find matching tasks

Able to find matching description of tasks using a command find <description>.

Added javadocs to Ui class.

BrendonLau committed:
Update code to align to coding standards

Main update is to update code to coding standard.

- Previously implemented checkstyle so will be tagging this commit for checkstyle.
- Update gradle to support javafx and run junit.

BrendonLau committed:
Update Duke's GUI

Added chat box UI.

YiFei2 commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,54 @@
+ * Represents commands entered by the input.
+ * Commands are to be processed by Parser.
+ */
+public enum Command {
```
Wow nice use of enum with HashMap to convert String to Command. Didn't think of that. 👍

YiFei2 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,72 @@
+/**
+ * This class handles errors in Duke.
+ */
+public class DukeException extends Exception {
```
I like how the exception handling is done! 👍 Does this mean that your program will only ever throw exceptions extending from this class?

BrendonLau replied:
Yes. I think it handles mainly the user input.

YiFei2 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,227 @@
+                output = ui.findMessage(found);
+                break;
+            default:
+                output = "Opps! something went wrong";
```
Perhaps you could have this in a Ui method to print for unexpected errors?

BrendonLau replied:
Yes. Thanks for pointing it out.

YiFei2 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,152 @@
+     * Create task based on task's type.
+     *
+     * @param description of the task.
+     * @param time        of the task for Deadline and Event.
+     * @param t           type of task.
```
Is the whitespace before each parameter description here is intended?

BrendonLau replied:
Yes, it is formatted by intelliJ. Are the white space not supposed to be there?

YiFei2 commented on src/main/java/duke/common/Const.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.time.format.DateTimeFormatter;
+
+public final class Const {
```
I like the use of another class to declare constants and to format date/time. I have not seen this before! 🤔 👍

clement0010 replied:
I think it would be better to use the full term for brevity?
```suggestion
public final class Constant {
```

YiFei2 reviewed (commented):
Really well done. I think I learnt a lot from reviewing your PR 😂.

clement0010 commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    public static Command getCommand(String c) {
+        // return null if command not in hashmap.
+        return commands.containsKey(c)
+                ? commands.get(c)
+                : null;
```
I think this part can simplify into a shorter version using `getOrDefault(c, null)`

BrendonLau replied:
Yep noted thanks!

clement0010 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,84 @@
+    private Ui ui;
+    private Parser parser;
+
+    private static final String DEFAULT_PATH = "data/duke.txt";
```
Love the fact that you declared a default path for the storage path.

clement0010 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,227 @@
+     * Create an instance of Parser
+     *
+     * @param taskList taskList to get information of tasks
+     * @param ui       to interact with the users after processing the input
```
Not sure if this whitespace is intended, perhaps it's an autoformatting issue?

BrendonLau replied:
Yes, I used the IntelliJ shortcut to format my code. Is the white space an issue?

clement0010 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,227 @@
+        taskList.markTask(stringToIndex(commandLine[1]));
+    }
+
+    private void parseUnMark(String[] commandLine) throws DukeException {
```
I think since `unmark`  is one word, the camel case version `parseUnmark` seems more appropriate?

BrendonLau replied:
You are right. Thanks for noticing.

clement0010 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,227 @@
+        }
+    }
+
+    private void parseList(String[] commandLine) throws DukeException {
```
Similar issue as `parseAdd`

clement0010 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,227 @@
+    private boolean isExit;
+
+    /**
+     * Create an instance of Parser
```
Small issue
```suggestion
     * Creates an instance of Parser
```

clement0010 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,227 @@
+        return taskList.find(commandLine[1]);
+    }
+
+    private void parseAdd(String[] commandLine) throws DukeException {
```
I think this function plays a `validator` role rather than a `parser`? Because it essentially is checking valid conditions for the `add` command.

BrendonLau replied:
The function also executes the task.. so it is more like an executeAdd?

clement0010 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,109 @@
+                taskList.addTask(taskString[2], Task.Type.TODO);
+                if (taskString[1].equals("1")) {
+                    taskList.markTask(idx);
+                }
```
Duplicated codes in line `95-97` and `101-103`, perhaps we could move this to after `line 105` to adopt the [DRY principle](https://dzone.com/articles/software-design-principles-dry-and-kiss)?

BrendonLau replied:
You are right. Thanks!

clement0010 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    /**
+     * Create task for Todo.
```
Small nitpick
```suggestion
     * Creates task for Todo.
```

clement0010 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    /**
+     * Mark a task as done.
```
Small nitpick
```suggestion
     * Marks a task as done.
```

clement0010 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    /**
+     * Find all the tasks that contains a matching description.
```
```suggestion
     * Finds all the tasks that contains a matching description.
```

clement0010 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    /**
+     * UnMark a task as not done.
```
```suggestion
     * Unmarks a task as not done.
```

clement0010 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,132 @@
+    private TaskList taskList;
+
+    /**
+     * Create an instance of Ui.
```
```suggestion
     * Creates an instance of Ui.
```

clement0010 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    /**
+     * Get the description of the task.
```
```suggestion
     * Gets the description of the task.
```

clement0010 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    /**
+     * Get the type of the task.
```
```suggestion
     * Gets the type of the task.
```

clement0010 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    /**
+     * Mark the task as done.
```
```suggestion
     * Marks the task as done.
```

clement0010 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    /**
+     * returns the list of tasks.
```
```suggestion
     * Returns the list of tasks.
```

clement0010 reviewed (commented):
Overall, no major violation of coding standards. Great job!

BrendonLau committed:
Update error handling for task marking

Handle the case when the task is already marked or unmarked

The user is able to mark or unable the same task multiple times

Adding exceptions to handle marking marked task to ensure logical interaction between user and the system

Let's add exceptions classes to handle marking and unmarking of task

BrendonLau committed:
Add assertions

Unnoticed bugs may cause unwanted errors and the program to crash

Assertions are used to ensure correct program state during run time

Let's
1.  Add assertions into classes Storage, TaskList and Tasks classes to handle assumptions of Task
2.  Refractor methods and variable for assertions

BrendonLau committed:
Merge pull request #1 from BrendonLau/branch-A-Assertions

Add assertions

BrendonLau committed:
Improve code quality

Some variable and method names can be improved for clarity
There are a few repeated on the knowledge of code used

Let's
1. Refractor similar uses of code into a function
2. Rename variables and methods to make it clear and concise
3. Improve code based on feedbacks from reviewers

BrendonLau committed:
Merge PRs

BrendonLau committed:
Merge pull request #2 from BrendonLau/branch-A-CodeQuality

Merging PRs

BrendonLau committed:
Update stream in matching tasks

BrendonLau committed:
Add reminder

No reminder in place to remind users of upcoming deadlines and events

Reminder is essential in quickly keep track of upcoming tasks that are important

Lets
- implement a reminder system to show events and deadlines that are not marked when user enter the application

BrendonLau committed:
Fix Junit test

Update test methods of Event and Deadline classes.

BrendonLau committed:
Update README.md

BrendonLau committed:
Update GUI to improve user interaction

GUI has the same dialog box for both input and output.

Changes in fxml file to make GUI more aesthetically pleasing and differentiate between input and output with different colors dialog box.

BrendonLau committed:
Update readme file

Adds the user guide for users.

BrendonLau committed:
Add product screenshot

Added Ui.png

BrendonLau committed:
Update README.md

BrendonLau committed:
Set theme jekyll-theme-midnight

BrendonLau committed:
Set theme jekyll-theme-cayman
